File: python/dsflix_sdk/modules/coins.py
from __future__ import annotations

from ..http import HttpClient


class CoinsModule:
    def __init__(self, http: HttpClient) -> None:
        self.http = http

    def get_balance(self) -> dict:
        return self.http.get("/api/v2/coins/balance")

    def transfer(self, recipient_display_id: str, amount: float, note: str | None = None) -> dict:
        return self.http.post("/api/v2/coins/transfer", {"recipient_display_id": recipient_display_id, "amount": amount, "note": note})

    def convert(self, reward_coins: float) -> dict:
        return self.http.post("/api/v2/coins/convert", {"reward_coins": reward_coins})

    def get_transactions(self, page: int | None = None, language: str | None = None, sort: str | None = None) -> dict:
        return self.http.get("/api/v2/coins/transactions", {"page": page, "language": language, "sort": sort})

File: src/modules/coins.ts
// ─────────────────────────────────────────────────────────────
//  DSFlix SDK  –  src/modules/coins.ts
// ─────────────────────────────────────────────────────────────

import { HttpClient } from '../http';
import {
  CoinBalance,
  CoinTransferParams,
  CoinTransferResponse,
  CoinConvertParams,
  CoinConvertResponse,
  TransactionHistoryResponse,
  TransactionHistoryOptions,
} from '../types';

export class CoinsModule {
  constructor(private http: HttpClient) {}

  /** Get the current user's coin and reward coin balance */
  getBalance(): Promise<CoinBalance> {
    return this.http.get('/api/v2/coins/balance');
  }

  /**
   * Transfer coins to another DSFlix user by their display ID.
   * Requires coins:write permission. Min 1 coin, max 10,000 per request.
   */
  transfer(recipientDisplayId: string, amount: number, note?: string): Promise<CoinTransferResponse> {
    const params: CoinTransferParams = { recipient_display_id: recipientDisplayId, amount, note };
    return this.http.post('/api/v2/coins/transfer', params);
  }

  /**
   * Convert reward coins to regular coins at a 2:1 ratio.
   * Amount must be even and ≥ 2.
   */
  convert(rewardCoins: number): Promise<CoinConvertResponse> {
    const params: CoinConvertParams = { reward_coins: rewardCoins };
    return this.http.post('/api/v2/coins/convert', params);
  }

  /** Get the user's coin transaction history */
  getTransactions(options?: TransactionHistoryOptions): Promise<TransactionHistoryResponse> {
    return this.http.get('/api/v2/coins/transactions', options as Record<string, string | number | boolean | undefined>);
  }
}
